LE Audio: CSIS Ordered Access procedure

Thalley opened the pull request:
Updates the API and implementation to implement the CSIP ordered access procedure. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/42341

Thalley committed:
Bluetooth: CSIS: Remove `const` for members in API

Having the members array being const causes a lot of
limitations in the implementatation, and was
occasionally even disregarded. Removed to make the
implementation more flexible.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Stored actives members by rank

Instead of getting doing look ups for each member
for the CSIS client procedures, we store the active
members ordered by rank.

This is slightly less efficient in cases of error,
but for some procedures as the upcoming Ordered
access procedure where we need to access each
member twice, ordering them once is more
efficient.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Implement ordered access procedure for CSIS client

the CSIS client now has a ordered access procedure function.
The function implements the procedure as per the CSIP spec,
where it will verify that 1 or more members are unlocked,
and the execture a procedure on each member in ascending
order of rank.

The procedure can be anything (even a non-BT procedure),
and it will be up to the applicaiton to implement
what to do in that.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

MariuszSkamra commented on include/bluetooth/audio/csis.h:
```diff
@@ -282,20 +282,22 @@ typedef void (*bt_csis_client_lock_changed_cb)(struct bt_csis_client_csis_inst *
- * @typedef bt_csis_client_lock_state_read_cb
- * @brief Callback for bt_csis_client_get_lock_state()
+ * @typedef bt_csis_client_ordered_access_cb_t
+ * @brief Callback for bt_csis_client_ordered_access()
```
If the callback is used from `bt_csis_client_ordered_access` context, why it's lifetime is different? I.e. why it's `bt_csis_client_cb` member?

Thalley replied:
Are you asking why this callback is not provided as an argument to `bt_csis_client_ordered_access` instead? 

That is certainly a possibility, but the `bt_csis_client_ordered_access_t` is more likely to change between each call to `bt_csis_client_ordered_access`, whereas `bt_csis_client_ordered_access_cb_t` is more likely the remain the same function. 

e.g. if you want to execute two procedures on an array of set members, you would call `bt_csis_client_ordered_access` with two different callbacks.

MariuszSkamra replied:
And then how do you know which procedure is complete?

Thalley replied:
Hmm, `bt_csis_client_ordered_access` could in theory be called with the same array of members and `set_info` at the same time (currently that would fail due to the `busy` flag), and then the callback would be called with the same parameters, yes. 

Currently (due to the `busy` flag) it is not a problem, but I see what you mean. 

Are you suggesting that we switch to use parameter-based callback for all client operations callbacks (like the GATT API)?

MariuszSkamra replied:
I guess that way the context would not be lost. From the API user perspective this would be used like this pseudo code:
```
static screen_indicator volume_indicator;

void group_volume_up_complete(group, err)
{
  screen_indicate(screen, screen_indicator, indicate_vol_up);
}

int group_volume_up_func(group/group_member)
{
   return volume_up(group_member);
}

int group_volume_up(group)
{
    return bt_csis_client_ordered_access(group, group_volume_up_func, group_volume_up_complete);
}
```

Thalley replied:
Right. 

I see what you mean, and I see the value of it, especially if we want to enable parallel operations (I don't recall the CSIP specifying that you can't do multiple ordered access procedures in parallel (of course, the "correct" way of doing this would be do do multiple things in the first callback, instead of doing the Ordered Access Procedure multiple times in parallel). 

However, the points you've made here is much more general than just this procedure, and generally applies to all client procedures.

MariuszSkamra replied:
I've created an issue here: https://github.com/zephyrproject-rtos/zephyr/issues/44867

MariuszSkamra commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -1412,26 +1559,31 @@ int bt_csis_client_lock(const struct bt_csis_client_set_member **members,
 		BT_DBG("Could not verify members: %d", err);
 		return err;
 	}
+	active_members_store_ordered(members, count, set_info, true);
```
```suggestion

	active_members_store_ordered(members, count, set_info, true);
```

new empty line

MariuszSkamra commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -1381,27 +1505,50 @@ int bt_csis_client_get_lock_state(const struct bt_csis_client_set_member **membe
 		BT_DBG("Could not verify members: %d", err);
 		return err;
 	}
+	active_members_store_ordered(members, count, set_info, true);
```
ditto

MariuszSkamra commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -140,6 +143,110 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
+	} else if (inst_1->cli.rank > inst_2->cli.rank) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int member_rank_compare_desc(const void *m1, const void *m2)
+{
+	struct bt_csis_client_set_member *member_1 = *(struct bt_csis_client_set_member **)m1;
+	struct bt_csis_client_set_member *member_2 = *(struct bt_csis_client_set_member **)m2;
+	struct bt_csis *inst_1;
+	struct bt_csis *inst_2;
+
+	/* TODO: lookup_instance_by_set_info should be able to work on a const
+	 * member, but it relies on `bt_conn_index` which doesn't take a const
+	 * bt_conn.
+	 */
+
+	inst_1 = lookup_instance_by_set_info(member_1, active.info);
+	inst_2 = lookup_instance_by_set_info(member_2, active.info);
+
+	if (inst_1 == NULL) {
+		BT_ERR("inst_1 was NULL for member %p", member_1);
+		return 0;
+	}
+
+	if (inst_2 == NULL) {
+		BT_ERR("inst_2 was NULL for member %p", member_2);
+		return 0;
+	}
+
+	if (inst_1->cli.rank < inst_2->cli.rank) {
+		return 1;
+	} else if (inst_1->cli.rank > inst_2->cli.rank) {
+		return -1;
+	}
+
+	return 0;
+}
```
Are those separate functions needed just to return -1 vs 1? Please extract common part

Thalley replied:
The only thing I can see that I can make common is moving the casting of `const void *` to `struct bt_csis_client_set_member *`, which would require me to create a function that takes a `const void *` that then casts this to a `struct bt_csis_client_set_member *` and then does the call to `lookup_instance_by_set_info`. 

But that extra function would just remove two lines 
```
	struct bt_csis_client_set_member *member_1 = *(struct bt_csis_client_set_member **)m1;
	struct bt_csis_client_set_member *member_2 = *(struct bt_csis_client_set_member **)m2;
```
and replace 
```c
	inst_1 = lookup_instance_by_set_info(member_1, active.info);
```
with something like
```c
	inst_1 = member_rank_compare_lookup_instance_by_set_info(m1, active.info);
```

I wanted to use `qsort_r` which allows for an additional argument, which could also make this a single function instead, but that is not POSIX compliant, and won't work with e.g. babblesim nor native_posix. 
The alternative is to store the comparison direction (ascending/descending) in the `active` struct. 

The last alternative is to create a new struct that holds a tuple of `struct bt_csis *` and return that. That way it could made common. 

If you want, I can create the tuple struct return type, and extract common code, or I can add a sorting direction variable in `active`, which would reduce this to a single function. Which do you prefer?

MariuszSkamra replied:
Try this one:
```
static int member_rank_compare_desc(const void *m1, const void *m2)
{
    return member_rank_compare_asc(m2, m1);
}
```

Thalley replied:
That's a great suggestion - Did not think of that myself :D

MariuszSkamra commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -140,6 +143,110 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
+
+	if (inst_1->cli.rank < inst_2->cli.rank) {
+		return 1;
+	} else if (inst_1->cli.rank > inst_2->cli.rank) {
```
else not needed as you return from if

Thalley replied:
The else covers the case where `inst_1->cli.rank == inst_2->cli.rank`

MariuszSkamra replied:
I know, I was thinking rather of:
``
if (a > b) {
  return 1;
}

if (a < b) {
  return -1;
}

return 0;
```

Thalley replied:
Right. 

Are you aware of any branching optimizations or downsides of doing
```c
if (a > b} {
  return 1;
}

if (b < a) {
  return -1;
}

return 0;
```

compared to 
```c
if (a > b} {
  return 1;
} else if (b < a) {
  return -1;
}

return 0;
```

I can change it no problem, even if it does not really change anything. 


With regards to doing `return inst_2->cli.rank - inst_1->cli.rank`, the `qsort` implementation does to provide enough information about whether the value can be less than -1 or greater than 1, but from the Linux `man`page:

> The comparison function must return an integer less than, equal to, or greater  than  zero if  the  first  argument  is considered to be respectively less than, equal to, or greater than the second.  If two members compare as equal, their order in the sorted array is  undefined.

So I guess I can just do the simple subtraction - Will update.

MariuszSkamra commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -140,6 +143,110 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
+			const uint8_t rank_1 = inst_1->cli.rank;
+			const uint8_t rank_2 = inst_2->cli.rank;
+
+			if (ascending) {
+				__ASSERT(rank_1 <= rank_2,
+					 "Members not sorted by ascending rank %u - %u",
+					 rank_1, rank_2);
+			} else {
+				__ASSERT(rank_1 >= rank_2,
+					"Members not sorted by descending rank %u - %u",
+					rank_1, rank_2);
```
Is this assert needed? Couldn't be verified by a bsim_test or unit test?

Thalley replied:
The application does have access to the rank value at this point, so it won't be able to do such as verification (actually tried to implement this in the babblesim tests first time around).

MariuszSkamra replied:
My point is that you sort the members, and then try to assert if not sorted. I'm OK with asserting things that one is not 100% certain, but here...

Thalley replied:
The solutions I see are
1) (current) Do an assert to verify that the sorting functionality works
2) Have a babblesim test use the internal header file and access the rank
3) Expose the rank to the application
4) Replace the asserts with a `BT_ERR`

MariuszSkamra replied:
If the check is necessary, I guess the assert is the best option from those you listed

MariuszSkamra reviewed (changes requested):
Some comments. Looking at the code I see that there is a static structure `active_members` that holds context. I guess the user might get into trouble when tries to use this API with multithreads and forbids concurrent operations on 2 sets.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/csis_client_test.c:
```diff
@@ -70,21 +71,18 @@ static void csis_lock_changed_cb(struct bt_csis_client_csis_inst *inst,
+		ordered_access_locked = true;
 	} else {
-		set_read_unlocked = true;
+		printk("Ordered access procedure finished\n");
```
`ordered_access_finished` was actually unused, so removed it.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/csis_client_test.c:
```diff
@@ -185,38 +196,40 @@ static void discover_members_timer_handler(struct k_work *work)
 	} else {
-		WAIT_FOR(set_read_unlocked);
+		WAIT_FOR(ordered_access_unlocked);
 	}
```
`ordered_access_finished` was actually unused, so removed it.

Thalley commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -1025,11 +1109,22 @@ static void csis_client_write_lock_cb(struct bt_conn *conn, uint8_t err,
-		int csis_client_err;
+		struct bt_csis_client_set_member *member;
 		struct bt_csis *prev_inst = cur_inst;
+		int csis_client_err;
 
-		cur_inst = get_next_inst_higher_rank(cur_inst->cli.rank,
-						     active.info);
+		member = active.members[active.members_handled];
+		cur_inst = lookup_instance_by_set_info(member, active.info);
+		if (cur_inst == NULL) {
+			BT_DBG("Failed to lookup instance by set_info %p",
+			       active.info);
+
+			active_members_reset();
+			if (csis_client_cbs != NULL &&
+			csis_client_cbs->lock_set != NULL) {
+				csis_client_cbs->lock_set(-ENOENT);
+			}
+		}
 
 		csis_client_err = csis_client_write_set_lock(cur_inst, true,
 							     csis_client_write_lock_cb);
```
I removed a bit of the duplicated code, but some still remain as that is not easily moved to a different function IMO.

Thalley commented on include/bluetooth/audio/csis.h:
```diff
@@ -327,21 +329,40 @@ bool bt_csis_client_is_set_member(uint8_t set_sirk[BT_CSIS_SET_SIRK_SIZE],
+ * @return true if the procedures can be successfully done, or false to stop the
+ *         procedure.
+ */
+typedef bool (*bt_csis_client_ordered_access_t)(const struct bt_csis_client_set_info *set_info,
+						struct bt_csis_client_set_member *members[],
+						size_t count);
```
I agree with all your points. 

> First of all, the bt_csis_client_ordered_access says that cb is called in the order of the member's ranks (if present) but the the callback does not have a member parameter, but the members (ordered by rank I suppose).
Thus, shouldn't this callback look like this ?

After a discussion with Asbjørn I changed it from a foreach to a single callback with all members. That seemed easier to use for all parties. 

> Secondly, you add another callback function named bt_csis_client_ordered_access_cb_t. That is confusing having those two with so similar names.

> Thirdly, Is it really needed to have 2 callbacks for this procedure? I have a feeling that the one given to bt_csis_client_ordered_access with additional err parameter (to indicate that procedure cannot be performed because one of the peer's is locked) should be sufficient.

One callback is called when the procedure is ready to be done on each member, and the other is when the ordered access procedure is completed. 

Ideally I'd like to reduce this to a single callback, but the parameters in each callback are sufficiently different that it doesn't really make sense IMO. 

I do, however, want to find a better name for these, as `_cb_` is the only difference between them... 

Basically we have
```c
bt_csis_client_ordered_access()
bt_csis_client_ordered_access_t() // callback
bt_csis_client_ordered_access_cb_t() // callback for bt_csis_client_ordered_access
```

Personally I think a new name for `bt_csis_client_ordered_access_t` is the best solution here. 

How do you feel about `bt_csis_client_ordered_access_ready_t` or `bt_csis_client_ordered_access_execute_t`? 
Do you have any other ideas, or perhaps an idea to reduce the callbacks to a single callback?

MariuszSkamra replied:
> After a discussion with Asbjørn I changed it from a foreach to a single callback with all members. That seemed easier to use for all parties.

OK, but the documentation was out of date I guess. Thanks for update ^^ Maybe it would be nice to have a info that `@ref bt_csis_client_ordered_access_cb_t` will be called when procedure is completed.

Thalley replied:
Added that in the documentation now as well :)

Thalley committed:
Bluetooth: CSIS: Client use `const` for conn lookup funcs

`bt_conn_index` has been changed to take a `const bt_conn`
which allows our functions that use this to also take a
`const` parameter.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Rebased to solve merge conflicts

MariuszSkamra commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -140,6 +143,90 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
+	      ascending ? member_rank_compare_asc : member_rank_compare_desc);
+
+	if (IS_ENABLED(CONFIG_ASSERT) &&
+	    IS_ENABLED(CONFIG_BT_DEBUG_CSIS_CLIENT)) {
```
I guess `CONFIG_BT_DEBUG_CSIS_CLIENT` is undesirable here, as when `CONFIG_ASSERT` is enabled you might want to assert, no matter if CSIS debug is enabled.

Thalley replied:
Good point. This was a leftover from when I just did a `BT_ERR` instead.

Thalley commented:
Fixes CI issue with array out of bounds (fixed by checking if `CONFIG_BT_MAX_CONN > 1`

larsgk commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -140,6 +143,90 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
+	return member_rank_compare_asc(m2, m1);
+}
+
+static void active_members_store_ordered(struct bt_csis_client_set_member *members[],
```
Looking through the PR, it doesn't seem like the original array of pointers is modified.  Is there a reason for this not being const?

Thalley replied:
Essentially it is because `bt_csis_client_set_member` contains the `bt_conn`, and all the lower layer APIs (including GATT) does not have the `bt_conn` parameter as const. 

Besides that, we also have that `bt_csis_client_set_member.insts[x].csis` that also need to be non-const as that will also change during the procedures. 

Some of this API is changed again in https://github.com/zephyrproject-rtos/zephyr/pull/45164 to accommodate the CAP initiator, so some of this will unfortunately be changed again.

larsgk replied:
ok, thanks - LGTM then

larsgk commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -140,6 +143,90 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
+	return member_rank_compare_asc(m2, m1);
+}
+
+static void active_members_store_ordered(struct bt_csis_client_set_member *members[],
```
Looking through the PR, it doesn't seem like the original array of pointers is modified.  Is there a reason for this not being const?

larsgk reviewed (approved):
LGTM

Thalley commented:
Rebased to solve merge conflicts

sjanc commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -137,6 +140,90 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
+		qsort(active.members, count, sizeof(members[0U]),
+		ascending ? member_rank_compare_asc : member_rank_compare_desc);
+
+		if (IS_ENABLED(CONFIG_ASSERT) && CONFIG_BT_MAX_CONN > 1) {
```
this is already underl BT_MAX_CON>1

asbjornsabo commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -183,43 +185,46 @@ static int member_rank_compare_asc(const void *m1, const void *m2)
-	/* TODO: lookup_instance_by_set_info should be able to work on a const
-	 * member, but it relies on `bt_conn_index` which doesn't take a const
-	 * bt_conn.
+	/* If we call the "compare ascinding" function with the members
```
ascinding -> ascending

Neat simplification, though.  Could this rather be done in the previous commit, where the functions are introduced?

asbjornsabo reviewed (changes requested):
Some minor things.
Also: There seems to be some overlap between the second and third commit - with some code introduced in the second commit being changed again in the third.  I think it would simplify the commits if the "net result" was introduced in the second commit.

Thalley commented on include/zephyr/bluetooth/audio/csis.h:
```diff
@@ -282,20 +282,22 @@ typedef void (*bt_csis_client_lock_changed_cb)(struct bt_csis_client_csis_inst *
- * set member is in the locked state, the remaining (if any) won't be read.
+ * If any of the set members supplied to bt_csis_client_ordered_access() is
+ * in the locked state, this will be called with @p locked true and @p member
+ * will the locked member, and the ordered access procedure is cancelled.
```
Added a `be`

Thalley commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -53,6 +53,8 @@ static struct active_members {
 	uint8_t members_count;
 	uint8_t members_handled;
 	uint8_t members_restored;
+
```
To separate the fields; `oap_cb` is not a field that represent the members or any of the counters, but rather a single value for the procedure itself.

Thalley commented:
@asbjornsabo w.r.t doing things in earlier commits: Yes, it is always possible to do the change in the previous commit, but it kind of goes against does a single thing in a commit (to keep commits as simple as possible). I'd rather keep it as is.

asbjornsabo commented:
> @asbjornsabo w.r.t doing things in earlier commits: Yes, it is always possible to do the change in the previous commit, but it kind of goes against does a single thing in a commit (to keep commits as simple as possible). I'd rather keep it as is.

I was thinking of the member_rank_compare_asc/desc - they were introduced in one commit, and then the "desc" was changed in the next one to just call the "asc". I think it would have been simpler to just introduce them in the final form in the first commit instead of spreading it over two.  IMO, that would be a single thing.
But I am OK to keep as is now.

asbjornsabo reviewed (approved):
My comments have been addressed.
